### PR TITLE
Improved unit test coverage and some refactoring

### PR DESCRIPTION
**Pull request by levelfivehub**
Hi,

* I added a gitignore on /coverage as it's quite neat to show coverage reports in the current directory.
* One of the classes was using the same error message copied and pasted.  I noticed one had an additional space so now it's consistent.  Should be easier to amend the error message if need be.  (Just incase we change a namespace as an example)
* I improved the unit test coverage

First contribution, hope it's well.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -7,6 +7,7 @@
 .*.un~
 nbproject
 tmp/
+coverage/
```
Comment (Maks3w) [inline .gitignore]: This is exclusive of your environment. Please undo.

File: src/Filter/FilterComposite.php [hunk 2]
```diff
@@ -10,6 +10,7 @@
 namespace Zend\Hydrator\Filter;
 
 use ArrayObject;
+use Zend\Hydrator\Filter\FilterInterface;
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Redundant import

File: src/Filter/FilterComposite.php [hunk 3]
```diff
@@ -48,8 +56,7 @@ public function __construct($orFilter = [], $andFilter = [])
             function ($value, $key) {
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Function could be extracted to local var and reused for both array_walks

Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Even better if `addFilter` its used

File: src/Filter/FilterComposite.php [hunk 4]
```diff
@@ -98,8 +104,7 @@ public function addFilter($name, $filter, $condition = self::CONDITION_OR)
     {
         if (!is_callable($filter) && !($filter instanceof FilterInterface)) {
             throw new InvalidArgumentException(
-                'The value of ' . $name . ' should be either a callable or ' .
-                'an instance of Zend\Stdlib\Hydrator\Filter\FilterInterface'
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Wrong instance type, not longer inherit from StdLib

File: test/Filter/FilterCompositeTest.php [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+use Zend\Hydrator\Filter\HasFilter;
+
+/**
+ * Unit tests for {@see FilterComposite}
+ *
+ * @covers \Zend\Hydrator\Filter\FilterComposite
+ */
+class FilterCompositeTest extends \PHPUnit_Framework_TestCase
+{
+    public function testOrFilterNotCallableOrFilter()
```
Comment (Maks3w) [inline test/Filter/FilterCompositeTest.php]: Please refactor all of these cases as data providers

File: test/ReflectionTest.php [hunk 6]
```diff
@@ -43,4 +43,20 @@ public function testCanHydrate()
 
         $this->assertSame($object, $this->hydrator->hydrate(['foo' => 'bar'], $object));
     }
+
+    public function testNotStringOrObjectOnExtract()
+    {
+        $this->setExpectedException(\InvalidArgumentException::class);
```
Comment (Maks3w) [inline test/ReflectionTest.php]: 1. Import exception class
2. Assert exception message too

File: test/ReflectionTest.php [hunk 7]
```diff
@@ -43,4 +43,20 @@ public function testCanHydrate()
...
+
+        $argument = (int) 1;
+        $this->hydrator->extract($argument);
+    }
+
+    public function testNotStringOrObjectOnHydrate()
+    {
+        $this->setExpectedException(\InvalidArgumentException::class);
```
Comment (Maks3w) [inline test/ReflectionTest.php]: 1. Import exception class
2. Assert exception message too

File: src/Filter/FilterComposite.php [hunk 8]
```diff
@@ -35,6 +36,13 @@ class FilterComposite implements FilterInterface
     const CONDITION_AND = 2;
 
     /**
+     * Error message for non callable variables
+     */
+    const ERROR_MESSAGE = <<<EOT
+The value of %s should be either a callable or an instance of Zend\Stdlib\Hydrator\Filter\FilterInterface
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: This smell bad, if same error message its used on different places then means its not enough informative about what is the source of the problem.

File: .gitignore (hunk 1 above)
Comment (levelfivehub) [inline .gitignore]: Will do.

File: src/Filter/FilterComposite.php (hunk 8 above)
Comment (levelfivehub) [inline src/Filter/FilterComposite.php]: Correct.  It's also not a good idea to have duplications.  Best to be efficient.  Recommendations?

Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Refactor the constructor for use addFilter method.

File: .gitignore (hunk 1 above)
Comment (levelfivehub) [inline .gitignore]: Removed from branch.

File: src/Filter/FilterComposite.php (hunk 4 above)
Comment (levelfivehub) [inline src/Filter/FilterComposite.php]: Where is it placed now?  Shall we just use FilterInterface::class to get the name?

File: src/Filter/FilterComposite.php (hunk 3 above)
Comment (levelfivehub) [inline src/Filter/FilterComposite.php]: Leave it with me.

File: test/ReflectionTest.php [hunk 9]
```diff
@@ -43,4 +44,20 @@ public function testCanHydrate()
 
         $this->assertSame($object, $this->hydrator->hydrate(['foo' => 'bar'], $object));
     }
+
+    public function testNotStringOrObjectOnExtract()
```
Comment (levelfivehub) [inline test/ReflectionTest.php]: Messages are now also tested.

File: test/Filter/FilterCompositeTest.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
+<?php
```
Comment (levelfivehub) [inline test/Filter/FilterCompositeTest.php]: Now using a dataProvider.

File: src/Filter/FilterComposite.php [hunk 11]
```diff
@@ -35,40 +35,22 @@ class FilterComposite implements FilterInterface
     const CONDITION_AND = 2;
 
     /**
-     * Define default Filter
+     * Error message for non callable variables
+     */
+    const ERROR_MESSAGE = <<<EOT
```
Comment (levelfivehub) [inline src/Filter/FilterComposite.php]: Need to know if the instance name is correct.

**Comment by levelfivehub** [comment]
Hi, does this look ok?

File: src/Filter/FilterComposite.php [hunk 12]
```diff
@@ -35,40 +35,22 @@ class FilterComposite implements FilterInterface
...
+The value of %s should be either a callable or an instance of Zend\Stdlib\Hydrator\Filter\FilterInterface
+EOT;
+
+    /**
+     * We can pass a list of OR/AND filters through construct
      *
-     * @param  array $orFilter
-     * @param  array $andFilter
-     * @throws InvalidArgumentException
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: This throws tag still been applicable

**Comment by levelfivehub** [comment]
@Maks3w  Can you please do a re-review.  After our conversation, I made the necessary changes as requested.  Test coverage is also higher than the original.

File: src/Filter/FilterComposite.php [hunk 13]
```diff
@@ -87,21 +65,15 @@ function ($value, $key) {
...
      * </code>
      *
-     * @param  string                   $name
-     * @param  callable|FilterInterface $filter
-     * @param  int                      $condition Can be either
-     *     FilterComposite::CONDITION_OR or FilterComposite::CONDITION_AND
-     * @throws InvalidArgumentException
-     * @return FilterComposite
+     * @param string $name
+     * @param FilterInterface $filter
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: Undo this changes

File: src/Filter/FilterComposite.php [hunk 14]
```diff
@@ -195,4 +167,25 @@ public function filter($property)
 
         return $returnValue;
     }
+
+    /**
+     * @param $name
+     * @param $filter
+     *
+     * @return bool
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: return void.

File: src/Filter/FilterComposite.php [hunk 15]
```diff
@@ -195,4 +167,25 @@ public function filter($property)
 
         return $returnValue;
     }
+
+    /**
+     * @param $name
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: complete types

File: src/Filter/FilterComposite.php (hunk 14 above)
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: add throws tag

**Comment by levelfivehub** [comment]
@Maks3w  Changes made, build running.

File: src/Filter/FilterComposite.php [hunk 16]
```diff
@@ -195,4 +169,24 @@ public function filter($property)
 
         return $returnValue;
     }
+
+    /**
+     * @param FilterInterface $filter
```
Comment (Maks3w) [inline src/Filter/FilterComposite.php]: `FilterInterface|callable`

**Comment by Maks3w** [comment]
Please squash your commits and ready to merge.

**Comment by levelfivehub** [comment]
@Maks3w  Squashed...

**Comment by levelfivehub** [comment]
@Maks3w  Thank you.
